### ISSUE TITLE
Compiler update and categories

### ISSUE THREAD

**Issue by adrienbernede** [LLNL/Umpire#287]
## Goal
I would like to update compilers/toolchains tested of LC systems, for example adding those used by RAJA.
However: 
 - Theres a lot of them, some probably useless.
 - There is no need to test all of them for each change, we could test all of them nightly.

That’s why I would like to classify them it two categories: core/main/basic and extension/secondary/advanced.

## Current state
On toss3 machines, the list of compilers/toolchains would be:

Umpire
- [ ] clang_3_9_1
- [ ] clang_4_0_0
- [ ] cudatoolkit_9_1
- [ ] gcc_4_9_3
- [ ] gcc_6_1_0
- [ ] gcc_7_1_0
- [ ] icpc_16_0_4
- [ ] icpc_17_0_2
- [ ] icpc_18_0_0
- [ ] pgi_17_10
- [ ] pgi_18_5

from RAJA
- [ ] clang_6_0_0
- [ ] clangcuda_6_0_0_nvcc_8_0
- [ ] gcc_7_3_0
- [ ] gcc_8_1_0
- [ ] icpc_18_0_2_gcc7headers
- [ ] icpc_19_0_4_gcc8headers

On blueOS machines this list currently is:

Umpire
- [ ] clang_3_9_1
- [ ] clang_4_0_0
- [ ] clang_coral_2017_06_29
- [ ] clang_coral_2017_08_31
- [ ] clang_coral_2017_09_06
- [ ] clang_coral_2017_09_18
- [ ] gcc_4_9_3
- [ ] nvcc_clang_coral_2017_06_29
- [ ] nvcc_clang_coral_2017_08_31
- [ ] nvcc_clang_coral_2017_09_06
- [ ] nvcc_clang_coral_2017_09_18
- [ ] nvcc_gcc_4_9_3
- [ ] nvcc_xl-beta-2017.09.13

from RAJA
- [ ] clang_6_0_0
- [ ] clang_coral_2018_08_08
- [ ] clang_upstream_2018_12_03
- [ ] clangcuda_upstream_2018_12_03_nvcc_9_2
- [ ] gcc_7_3_1
- [ ] nvcc_10_clang_6_0_0
- [ ] nvcc_10_gcc_7_3_1
- [ ] nvcc_10_xl_2019_02_07
- [ ] nvcc_10_xl_2019_04_19
- [ ] nvcc_10_xl_2019_06_12
- [ ] nvcc_9_1_clang_coral_2018_08_08
- [ ] nvcc_9_2_clang_6_0_0
- [ ] nvcc_9_2_clang_coral_2018_08_08
- [ ] nvcc_9_2_clang_upstream_2018_12_03
- [ ] nvcc_9_2_gcc_7_3_1
- [ ] nvcc_9_2_xl_2019_02_07
- [ ] nvcc_9_2_xl_2019_04_19
- [ ] nvcc_9_2_xl_2019_06_12
- [ ] pgi_19_7
- [ ] xl_2019_02_07
- [ ] xl_2019_04_19
- [ ] xl_2019_06_12

## Additional considerations:
- RAJA toolchains use different CUDA versions, which I would like to do with Umpire as well.
- The selection could also serve to identify which toolchains we would like to test for Spack packaging.

## Sum-up:
I can add all toolchains fairly easily. But I need a selection that would be used on every PR.
Then we should remove unnecessary toolchains.

**Comment by rhornung67** [comment]
@adrienbernede I don't think we need to run tests with all the compiler you listed. We maintain a bunch of build scripts in RAJA to make it easy to test different compiler versions. I think it suffices to pick a couple of relevant versions of each compiler (based on what our application users use), especially since the blueos clang and xl compilers are still being updated frequently. For example, AFAIK, everything IBM did in the clang-coral compiler has been upstreamed into the main clang compiler. So I don't think it's worth it to test many of the older ones unless we know folks are using them.

In other words, we can prune this list way down.  Also, it looks like you are looking at an older version of RAJA. I did house cleaning and updating in the last release.

**Comment by adrienbernede** [comment]
@rhornung67 I agree with you that this list can be cut down, at least for the most frequent tests.
I am simply asking for a concrete list of compiler which should be tested on a systematic manner.
Then, I don’t see any problem in testing a larger set of compilers from time to time (typically a scheduled check of master branch).

As you said, it depends on users, on the project, on the machine, etc. So I don’t think I can come up with a relevant proposal myself, which is why I am asking for help.

I haven’t updated RAJA in a while that’s true.

**Comment by davidbeckingsale** [comment]
Hey @adrienbernede  - thanks for compiling the list. I think it might be appropriate to run our tests on the latest version of each compiler, and then have a weekly test that runs the older versions.

What do you think @rhornung67 ?

**Comment by rhornung67** [comment]
Makes sense.

**Comment by adrienbernede** [comment]
OK!
